Source/Evolve/WarpXEvolve.cpp

```
@@ -192,7 +192,8 @@ WarpX::Evolve (int numsteps)
 
         // Update timestep for electrostatic solver if a constant dt is not provided
         // This first synchronizes the position and velocity before setting the new timestep
-        if (electromagnetic_solver_id == ElectromagneticSolverAlgo::None &&
+        if ((electromagnetic_solver_id == ElectromagneticSolverAlgo::None ||
+            evolve_scheme == EvolveScheme::ThetaImplicitEM) &&
             !m_const_dt.has_value() && m_dt_update_interval.contains(step+1)) {
```

is `!m_const_dt.has_value()` needed here?

---

This is the full code:
https://github.com/BLAST-WarpX/warpx/blob/2fd719c953ff281b50f9f4b823703e66c4aa37d1/Source/Evolve/WarpXEvolve.cpp#L193-L202

Based on the comments and the names of the functions it seems to me that it is correct as is. It seems that we want to update the time step by calling `UpdateDtFromParticleSpeeds` only if the time step is not required to be constant (i.e., only if `m_const_dt.has_value()` returns `false`).

Why would we allow to update the time step if the user requested a constant time step in the inputs file?

This was added by @archermarx in #5176, if I'm not mistaken.

---

_Why would we allow to update the time step if the user requested a constant time step in the inputs file?_

- I guess it's just a matter or what takes priority assuming both are defined in the input deck. If the m_dt_update_interval is defined, then it could take priority and not care if you have const_dt defined too.

---

Ah, I didn't understand that they can both be specified in the inputs. If they're mutually exclusive then maybe we should raise an error if a user specifies both and instruct the user to choose what they want, a constant time step or a time step that can be updated. I don't have in mind the use cases, so my suggestion may be off.